include/swift/IDE/SourceEntityWalker.h

```diff
@@ -88,6 +88,15 @@ class SourceEntityWalker {
   /// false, the remaining traversal is terminated and returns failure.
   virtual bool walkToDeclPost(Decl *D) { return true; }
 
+  /// This method is called immediately before \c walkToDeclPre. It is guranteed
+  /// to be called in balance with its counterpart \c endBalancedDeclVisit even
+  /// if \c walkToDeclPre returns false.
+  virtual void beginBalancedDeclVisit(Decl *D){};
+
+  /// This method is called immediately after \c walkToDeclPost or after \c
+  /// walkToDeclPre if it returns false.
+  virtual void endBalancedDeclVisit(Decl *D){};
```

Comment:
guranteed  -> guaranteed

@benlangmuir and @rintaro do you have any ideas for naming these? They're not *just* balanced but will have a separate order to `walkTo*` and may be called on more decls than `walkToDecl*`.

Ie. The `TypeRepr`/`Expr`s in custom attributes will be walked before `walkToDecl*` for the attributed decl, but between `*BalancedDeclVisit`.

And another example is if `shouldWalkInactiveConfigRegion` is set then `*BalancedDeclVisit` will be called for an `IfConfigDecl` but `walkToDecl*` will not.

Reply:
`*LookaheadDeclVisit` perhaps?

Reply:
> They're not just balanced but will have a separate order to walkTo* and may be called on more decls than walkToDecl*.
> Ie. The TypeRepr/Exprs in custom attributes will be walked before walkToDecl* for the attributed decl, but between *BalancedDeclVisit.
> And another example is if shouldWalkInactiveConfigRegion is set then *BalancedDeclVisit will be called for an IfConfigDecl but walkToDecl* will not.

Interesting, I hadn't noticed this.  There may be good reasons for having a difference in what decls will get the callbacks, but we should be able to document what the difference is.  I do think we need to eliminate the differences in walk _order_ though - that just seems like a foot gun.

Reply:
>  I do think we need to eliminate the differences in walk order though - that just seems like a foot gun.

It seems to me (from a newcomers perspective), that perhaps `IndexSwiftASTWalker` has outgrown the [stated design goals](https://github.com/apple/swift/blob/main/include/swift/IDE/SourceEntityWalker.h#L43-L45) of `SemaAnnotator/SourceEntityWalker`. Due to the introduction of property wrappers, simply walking in source order is no longer practical, at least in order to achieve the index structure this PR introduces. It also seems `IndexSwiftASTWalker` and `SourceEntityWalker` have diverged in regards to implicit declarations, `IndexSwiftASTWalker` does now index some implicit declarations (constructors, accessors). I don't really see an easy solution to this, other than a huge refactor to decouple `IndexSwiftASTWalker` and `SourceEntityWalker`.

Reply:
Another instance of friction is custom attribute initializers. `SemaAnnotator::handleCustomAttributes` [explicitly ignores](https://github.com/apple/swift/blob/main/lib/IDE/SourceEntityWalker.cpp#L644) implicit initializers, whereas `IndexSwiftASTWalker` [explicitly adds them](https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L343).

Reply:
> `IndexSwiftASTWalker` does now index some implicit declarations (constructors, accessors)

I believe implicit constructors are walked by `SourceEntityWalker` as well (see `SemaAnnotator::shouldIgnore`).

> Another instance of friction is custom attribute initializers. `SemaAnnotator` explicitly ignores implicit initializers, whereas `IndexSwiftASTWalker` explicitly adds them.

In general I think we probably just want to add an option to `SemaAnnotator` that walks more implicit nodes, rather than customise the behaviour in the subclasses. Though I'm not sure if anything *wouldn't* want to walk these cases.

Regardless, not something to tackle in this PR. Some name that makes it clear that it's visiting in the ASTWalker order versus source order and a clear comment seems reasonable to me. Another alternative could be to implement all the container logic in `SourceEntityWalker`, hide these new methods (private + friend SemaAnnotator), and move `getContainer` and `getParent` to `SourceEntityWalker`.

Reply:
> Some name that makes it clear that it's visiting in the ASTWalker order versus source order and a clear comment seems reasonable to me. Another alternative could be to implement all the container logic in SourceEntityWalker, hide these new methods (private + friend SemaAnnotator), and move getContainer and getParent to SourceEntityWalker.

Okay I could accept either of these options, I think.

Reply:
OK, I've improved the documentation for now. Pushing all this down into `SourceEntityWalker` sounds like a good idea though, I'll do that in a follow-up PR.